pkg/networkserver/redis/registry.go

```diff
@@ -480,305 +536,110 @@ func (r *DeviceRegistry) RangeByUplinkMatches(ctx context.Context, up *ttnpb.Upl
-			fieldKeyCurrent,
-			matchUIDKeyCurrentLE,
-			matchUIDKeyCurrentGT,
-			matchFieldKeyCurrent,
-		}
-	} else {
-		matchKeys = []string{
-			matchResultKey,
-
-			addrKeyCurrent,
-			fieldKeyCurrent,
-			matchUIDKeyCurrentLE,
-			matchUIDKeyCurrentGT,
-			matchFieldKeyCurrent,
-
-			addrKeyPending,
-			fieldKeyPending,
-			matchUIDKeyPending,
-			matchFieldKeyPending,
+	var (
+		currentSessionScoresCmd  *redis.ZSliceCmd
+		currentSessionMappingCmd *redis.StringStringMapCmd
+		pendingSessionScoresCmd  *redis.ZSliceCmd
+		pendingSessionMappingCmd *redis.StringStringMapCmd
+	)
+	if _, err := r.Redis.TxPipelined(ctx, func(p redis.Pipeliner) error {
+		currentSessionScoresCmd = ZRangeArgsWithScores(ctx, p, addrKeyCurrent)
+		currentSessionMappingCmd = p.HGetAll(ctx, fieldKeyCurrent)
+		if !ackFlag {
+			pendingSessionScoresCmd = ZRangeArgsWithScores(ctx, p, addrKeyPending)
+			pendingSessionMappingCmd = p.HGetAll(ctx, fieldKeyPending)
 		}
+		return nil
+	}); err != nil {
+		return ttnredis.ConvertError(err)
 	}
 
-	trace.Log(ctx, "ns:redis", "run deviceMatchScript")
-	vs, err := ttnredis.RunInterfaceSliceScript(ctx, r.Redis, deviceMatchScript, matchKeys, lsb, cacheTTL.Milliseconds()).Result()
+	var (
+		currentSessionSet []sessionEntry
+		pendingSessionSet []sessionEntry
+		err               error
+	)
+	currentSessionSet, err = buildSortedSessionSet(currentSessionScoresCmd, currentSessionMappingCmd, true, lsb)
 	if err != nil {
-		if err == redis.Nil {
-			return errNoUplinkMatch.WithCause(errNoCandidates)
-		}
-		return ttnredis.ConvertError(err)
+		return err
 	}
-	if len(vs) < 1 {
-		panic("empty match script return value")
+	if !ackFlag {
+		pendingSessionSet, err = buildSortedSessionSet(pendingSessionScoresCmd, pendingSessionMappingCmd, false, 0)
+		if err != nil {
+			return err
+		}
 	}
-	matchType, ok := vs[0].(string)
-	if !ok {
-		panic(fmt.Sprintf("expected first match script return value element to be a string, got '%v'(%T)", vs[0], vs[0]))
+
+	fillContext := func(ctx context.Context, uid string) (context.Context, *ttnpb.EndDeviceIdentifiers, error) {
+		ids, err := unique.ToDeviceID(uid)
+		if err != nil {
+			log.FromContext(ctx).WithError(err).Error("Failed to parse UID as device identifiers")
+			return ctx, nil, errDatabaseCorruption.WithCause(err)
+		}
+		return ctx, &ids, nil
 	}
-	processResult := func(ctx context.Context, s string) error {
-		defer trace.StartRegion(ctx, "process result").End()
 
-		if s == string(noUplinkMatchMarker) {
-			return errNoUplinkMatch.WithCause(errNoMatchMarker)
+	for _, session := range currentSessionSet {
+		ctx, ids, err := fillContext(ctx, session.UID)
+		if err != nil {
+			return err
 		}
-		ctx = log.NewContextWithField(ctx, "match_key", matchResultKey)
-		res := &UplinkMatchResult{}
-		if err := msgpack.Unmarshal([]byte(s), res); err != nil {
-			log.FromContext(ctx).WithError(err).Error("Failed to unmarshal match result")
-			return errDatabaseCorruption.WithCause(err)
+		if ids == nil {
```

Comment:
When is this happening?

Reply:
In Enterprise, when you have traffic coming from tenant `t1`, you don't want to match end devices in tenant `t2` unless the traffic was routed via PB.

https://github.com/TheThingsIndustries/lorawan-stack/blob/afb7ca7dd431b6a8447ef5c1078307efcc1abf27/pkg/networkserver/redis/registry.go#L710-L716

`t1` being the caller (in the context of a gateway), `t2` being the one who did register the possible match.

Reply:
See https://github.com/TheThingsIndustries/lorawan-stack/blob/afb7ca7dd431b6a8447ef5c1078307efcc1abf27/pkg/networkserver/redis/registry.go#L710-L716 . We basically ensure that we don't match the traffic without the packet broker reflecting it to the instance. The situation when this would happen when two tenants share the same cluster, so both are going to have the session registered under the packet broker tenant. When `t1` sees the traffic, we don't want to match it in `t2` unless PB actually sent it back.